app/referee/tasks.py

```diff
@@ -57,18 +57,62 @@ def run_and_score_submission(client, submission):
 def submit_submission_run(client, submission_run):
     delayed_conditional = dask.delayed(True)
     delayed_conditional = _run(submission_run, delayed_conditional)
+    if settings.VISUALIZE_DASK_GRAPH:
+        delayed_conditional.visualize(filename="task_graph.svg")
     future = client.submit(delayed_conditional.compute)  # pylint:disable=no-member
     logger.info("Future key: %s", future.key)
 
     dd.fire_and_forget(future)
     return future
 
 
+@dask.delayed(pure=False)
+def cache_containers(submission_run, delayed_conditional):
+
+    if not delayed_conditional:
+        return 1
+
+    container = submission_run.submission.container
+    aws_login_func = (
+        utils.get_aws_credential_function(container.uri)
+        if settings.LOGIN_TO_AWS
+        else None
+    )
+    submission_run.append(stdout=f"{container.uri}\n")
+    submission_run.append(stdout=f"{container.container_type}\n")
+
+    if settings.CONTAINER_FILES_ROOT:
+        container_save_path = (
+            settings.CONTAINER_FILES_ROOT
+            / f"{container.label.replace('/', '_')}_{container.tag}.sif"
+            if container.tag
+            else settings.CONTAINER_FILES_ROOT
+            / f"{container.label.replace('/', '_')}.sif"
+        )
+    else:
+        container_save_path = None
+
+    pull_code, stdout, stderr = ever_given.wrapper.pull_container(
+        container.uri,
+        container.container_type,
+        settings.CONTAINER_ENGINE,
+        container_save_path,
+        aws_login_func,
+    )
+    submission_run.append(stderr=stderr)
+    submission_run.append(stdout=stdout)
+    submission_run.append(stdout=f"PULL EXIT CODE: {pull_code} {type(pull_code)}\n")
+    return pull_code
+
+
 def _run(submission_run, delayed_conditional):
+
+    pull_code = cache_containers(submission_run, delayed_conditional)
+
     if submission_run.submission.challenge.current_batch_group() is None:
-        statuses = _run_evaluations(submission_run, delayed_conditional)
+        statuses = _run_evaluations(submission_run, pull_code, delayed_conditional)
```

Comment:
We have both delayed_conditional and pull_code that are doing essentially the same thing. Do we need two arguments or can we change how we use delayed_conditional?

Reply:
Perhaps a more sophisticated conditionals object would handle this better. TBD